### PR TITLE
feat(Dropdown): add category support

### PR DESCRIPTION
**Pull request by ashharrison90**
for https://github.com/carbon-design-system/carbon/issues/2595

Have added a new optional function property `itemToCategory` to `Dropdown`. This allows someone to categorise items based on a particular item property.

This is how it looks when categories are used:
image

Otherwise the look is unchanged.

Updated + added some new snapshot tests in both `ListBox` and `Dropdown`, and added a new story when using categories.

#### Changelog

**New**

- `ListBoxMenuCategory` - new type of menu component for category headers + associated css

**Changed**

- `Dropdown` - new optional property `itemToCategory`

#### Testing / Reviewing

- Pull it down and have a play around!

### CODE REVIEW

**Review by joshblack — COMMENTED** [review]
Hi @ashharrison90! 👋 Thanks so much for taking the time to put this together 🙏 Real quick, how were you anticipating a screen reader user to leverage this control? Wanted to make sure I understood the interaction pattern before testing 👍

**Comment by ashharrison90** [comment]
Hi @joshblack. I'm not sure I understand the question exactly. I'd expect a screen reader to read the category header like it would read any other text on a page. Would a screen reader work better if the items were nested under a category? 🤔

**Comment by joshblack** [comment]
Totally fair! When using VoiceOver on Safari, unfortunately it seems like the categories themselves aren't being announced which is why I was checking to see. Quick demo:

demo

**Comment by ashharrison90** [comment]
So I've tried providing every alternative in the world for this and it never seems to work. I've tried nesting the options within the categories, I've tried providing alternative labels to each item in the form of `aria-label="Option 1 (Europe)"`. Nothing seems to change what VoiceOver reads out.

Any ideas? Do you guys consider this a blocker?

File: packages/react/src/components/Dropdown/Dropdown.js
Comment (emyarod) [inline packages/react/src/components/Dropdown/Dropdown.js]: if I'm not mistaken we can just `return Object.values(categoryMap).flat()` and remove `const result = []` above as well

File: packages/react/src/components/Dropdown/Dropdown.js
Comment (emyarod) [inline packages/react/src/components/Dropdown/Dropdown.js]: small nitpick but the `return` statement can go outside of the `else` block right?

File: packages/components/src/components/list-box/_list-box.scss
Comment (emyarod) [inline packages/components/src/components/list-box/_list-box.scss:514]: might need confirmation from design but I believe the border here should become transparent when hovering over a dropdown item above a category

image

File: packages/react/src/components/Dropdown/Dropdown.js
Comment (ashharrison90) [reply packages/react/src/components/Dropdown/Dropdown.js]: yep, that's much nicer 👍

File: packages/react/src/components/Dropdown/Dropdown.js
Comment (ashharrison90) [reply packages/react/src/components/Dropdown/Dropdown.js]: yep 👍

File: packages/components/src/components/list-box/_list-box.scss
Comment (ashharrison90) [reply packages/components/src/components/list-box/_list-box.scss:514]: i'll wait for confirmation from design on that - personally i think being able to see the line splitting the categories helps a little 🤷‍♂thanks!

**Comment by joshblack** [comment]
@ashharrison90 unfortunately yeah we need this control to pass AVT3. The component itself is being worked on to address most of our AVT3 violations in our current accessibility project so part of this is 100% on us, as well!

**Comment by abbeyhrt** [comment]
@ashharrison90 The screenreader stuff is an incredibly frustrating problem to work on, thank you for taking the time to look into it! I'm on the a11y team for Carbon and right now, I don't think we can ever get the categories to be read to screenreaders because of deep problems intrinsic to Dropdown 😞. A11y-wise, Dropdown is being held together with duct tape and barely passes, that's why any added aria-labels just won't announce. We're currently working on fixing these problems with our Dropdown and we would love to have this enhancement after those problems are addressed! Would you mind if we closed this PR for now and reopen it once it can be added accessibly? Once that's done, we would love to work with you to get this added in!

**Comment by ashharrison90** [comment]
hey @abbeyhrt @joshblack 

so our product is currently making a concerted effort to move to Carbon 10 by the end of the year. as mentioned in the linked issue, this is a real blocker as we need category support within `Dropdown` to maintain existing functionality. 

is there any timescale on fixing those problems? or is there anything i can do to help? happy to do some more digging if someone points me in the right direction. 🤓

**Comment by abbeyhrt** [comment]
@ashharrison90 Depending on the scope of the changes that need to be made, a fix could come in one our next two minor versions, so right after the New Year at the latest.  In the meantime, it's not as pretty but the `Select` component does currently support categories if you would be able to switch to that one?

**Comment by aagonzales** [comment]
Is there a design spec or design file for this? (besides the one image) I'd like to take a look at it.

**Comment by ashharrison90** [comment]
I don't think our designers would ever be happy with switching to `Select` unfortunately 😅

@aagonzales I don't think there's a spec file for this, I was going off your designs in the original issue linked above. 

My job atm is basically moving to Carbon 10, so if there's any way I can help move this along I'm happy to.

**Comment by aagonzales** [comment]
I'm going to do some quick visual studies just to make sure this is the correct direction for with the v10 styles.

**Review by aagonzales — COMMENTED** [review]
Ok this is the correct direction, just a few small tweaks:
- the last item in a category list should have `12px` of padding below it before the divider rule.
- the color of the category label should should use `$text-05`

image

**Comment by asudoh** [comment]
Weighing in; You can monitor the dropdown items being announced by going to Console tab in Chrome DevTools, clicking on "eye" icon at the top and put the following:

```javascript
(() => {
  const iframe = document.querySelector('iframe');
  const doc = iframe ? iframe.contentDocument : document;
  return doc.querySelector('[aria-live]').textContent;
})()
```

Not sure if the PR author wants to figure out a way to influence that upon navigating (or introduce an similar element separately), but if it happens, we may have a chance to make some progress here.

**Comment by ashharrison90** [comment]
thanks for the nudge in the right direction @asudoh 🤓 

i've added a new custom implementation of `getA11yStatusMessage` to be passed in to `Downshift`. this now appends categories to to the `aria-label` for each item, which i believe is an accepted pattern for screen readers.

i've also addressed the small design tweaks @aagonzales 🙌 

let me know if there's anything else to do 👍

File: packages/components/src/components/list-box/_list-box.scss
Comment (jnm2377) [reply packages/components/src/components/list-box/_list-box.scss:514]: @laurenmrice should this hover still have the bottom border?

Comment (ashharrison90) [reply packages/components/src/components/list-box/_list-box.scss:514]: @jnm2377 @laurenmrice if you look at an updated design here: https://github.com/carbon-design-system/carbon/pull/4449#pullrequestreview-307455042 you'll see it's much more necessary now 😛

**Review by aagonzales — COMMENTED** [review]
Looks good! Just need to add the extra padding to the last item in a category.  

image

**Comment by ashharrison90** [comment]
fixed the padding now

have also restructured slightly to nest `MenuItem`s within each `MenuCategory` in the DOM

**Comment by asudoh** [comment]
Thank you for your update @ashharrison90 - Looks basically good, need to see if our existing code makes the content being announced translatable or not. One question; Does your new code take the whole control over the announced content? If so, it'd be great if we can make it translatable.

@dakahn Would you have a chance to see if @ashharrison90's solution looks good for your a11y perspective? Thanks!

**Comment by ashharrison90** [comment]
@asudoh it uses the existing `itemToString` and the new `itemToCategory` functions, so i think it's fine as we leave it up to a consumer to do the translation in both cases. and yes it takes full control over the announced content as it's passed in to `Downshift`

File: packages/react/src/components/Dropdown/Dropdown.js
Comment (asudoh) [inline packages/react/src/components/Dropdown/Dropdown.js]: Can you make this translatable, with [`translateWithId` interface](https://github.com/carbon-design-system/carbon/blob/v10.7.0/packages/react/src/components/Dropdown/Dropdown.js#L94-L97)? Ref: https://github.com/carbon-design-system/carbon/blob/v10.7.0/packages/react/docs/g11n.md

File: packages/react/src/components/Dropdown/Dropdown.js
Comment (asudoh) [inline packages/react/src/components/Dropdown/Dropdown.js]: Can you make this translatable, with [`translateWithId` interface](https://github.com/carbon-design-system/carbon/blob/v10.7.0/packages/react/src/components/Dropdown/Dropdown.js#L94-L97)? Ref: https://github.com/carbon-design-system/carbon/blob/v10.7.0/packages/react/docs/g11n.md

**Review by asudoh — COMMENTED** [review]
Thank you for clarifying @ashharrison90!

File: packages/react/src/components/Dropdown/Dropdown.js
Comment (ashharrison90) [reply packages/react/src/components/Dropdown/Dropdown.js]: i think that's beyond the scope of this PR - that content was never being translated before. if someone who knows more about the `translateWithId` interface wants to follow up to translate the strings, it's at least possible now

File: packages/react/src/components/Dropdown/Dropdown.js
Comment (ashharrison90) [reply packages/react/src/components/Dropdown/Dropdown.js]: see other comment

**Review by abbeyhrt — CHANGES_REQUESTED** [review]
@ashharrison90 this definitely is a step in the right direction accessibility-wise and I appreciate you doing this work! Right now though, this isn't how we would want the categories to announce. We want them to announce the way that the `Single select with option groups` announces in VoiceOver and JAWS from here: https://dequelabs.github.io/combobo/demo/. In JAWS, which is our main screenreader requirement, it is reading "Option 1 left paren Europe right paren", which is also not ideal. If this can copy the example from deque, then I would consider this good to go in terms of AVT3.

File: packages/react/src/components/Dropdown/Dropdown.js
Comment (asudoh) [reply packages/react/src/components/Dropdown/Dropdown.js]: OK fair enough - Thank you for clarifying.

**Comment by ashharrison90** [comment]
ok i've improved the accessibility message to announce when it enters a new group + the length of that group, and also fixed a bug i didn't spot before where the indexing was all messed up due to the presence of categories 👍

**Comment by joshblack** [comment]
We're unfortunately actively running into issues with downshift which is what is fundamentally causing a lot of these issues. Part of the work to address existing issues on the carbon side will make this easier.

With respect to the implementation, I think to match the example from deque we'd need to make sure that the groups are wrapped in `role="group"` with `aria-labelledby` pointing to an id for the visual node for the category item. Was trying to see if downshift had a solid example for this but couldn't find one that hits WCAG 2.0 AA

With respect to the API, it definitely is rough. I think this really shows the shortcomings of our API changes, unfortunately 🤦‍♂ (definitely on me here). I think we'll run into issues having this work with ListBox.{Menu,MenuItem}, as well. Would be worth investing the time with the group tackling the internal updates for this component to see if it matches what our plans are. Most likely, this will mean removing downshift to try and simplify how we are putting things together.

Just want to hop in and add more context here, would hate for you all to spend time on this only to have the implementation be swapped out. Always happy to chat on slack or in a webex, feel free to reach out on Slack `@josh` 👍

**Comment by ashharrison90** [comment]
i can put the `role` and `aria-label`s etc in,  it just won't do anything because Downshift overrides it all with its `getA11yStatusMessage` function

i'll reach out on slack 👌

**Comment by ashharrison90** [comment]
i've added the correct `role`s and `aria-label`s etc, such that the accessibility tree looks to be working properly now. as i said, this doesn't affect what a screen reader says at all, but i've put in some separate code to deal with that.

another thing mentioned was changes to the api, and if we could avoid them. as far as i can see, the only external api change is to add `itemToCategory` to `Dropdown`, and i'm not sure how any implementation of categories (now or in a future refactored version of `Dropdown`) could avoid having some extra prop to identify which property in an item denotes it’s category.

open to suggestions tho, so lemme know 😃

**Review by aagonzales — APPROVED** [review]
Looks good from a design standpoint.

**Comment by asudoh** [comment]
@abbeyhrt Does the update to the a11y stuff look good? Thanks!

**Comment by abbeyhrt** [comment]
hey @ashharrison90! I checked VoiceOver and that sounds good to me! It does seem like the changes you made have caused some DAP violations though, would you be able to address those? To recreate it, you'll have to test DAP when the Dropdown is open. You might already know this but you can ignore the "All content must reside within a HTML5..." violation, it's coming from our storybook implementation 👍 
Screen Shot 2019-11-06 at 10 55 39 AM

**Comment by elizabethsjudd** [comment]
@abbeyhrt the active descendant violation might be related to the bug that I just fixed around that attribute: https://github.com/carbon-design-system/carbon/pull/4501

**Comment by abbeyhrt** [comment]
@elizabethsjudd thanks, I'll look into it!

**Comment by abbeyhrt** [comment]
Sorry to go back on this but please ignore my previous comment about the VoiceOver, I'm not sure if it has to do with merging in master but I tested it again and the categories are no longer reading correctly. The category name is being announced but it's not announcing the number of options in each category group. I can also select the category name and it closes the dropdown, so it doesn't seem to be disabled fully? You can ignore what I said about the DAP violations though, those are present in all versions of the Dropdown.

**Comment by ashharrison90** [comment]
Sorry @abbeyhrt, I can't recreate either of those problems (you can't see my cursor, but I tried clicking the category headers and nothing happens):

Nov-07-2019 12-01-55

**Comment by asudoh** [comment]
@abbeyhrt Would you want to elaborate more on your finding so @ashharrison90 can re-create? Thanks!

**Comment by joshblack** [comment]
Revisiting this, going to close this out as we chatted a couple weeks back about not pursuing this without addressing other a11y violations in these components. Just wanted to say thanks so much for all your work here and willingness to work with our requirements, and just offer an apology that it didn't work out for your timeline. I am so sorry that it didn't work out 😞 

Moving forward, would love to connect and see how we can loop back for this requirement. Feel free to reach out on Slack if you feel like this is in error and I'm happy to chat about it!

**Comment by ashharrison90** [comment]
image

just to clarify, we last had a chat over a month ago. here's the end of the conversation. i've addressed every a11y violation that was mentioned to me, including some that already existed before this PR, so i've no idea what other a11y violations you want me to address. 

i cannot recreate https://github.com/carbon-design-system/carbon/pull/4449#issuecomment-550459850, and @abbeyhrt has helpfully not replied in a month either.

why even make this open source if you don't want any contributions? 

i look forward to your non-violating implementation of dropdown coming soon ™️ 😴

**Comment by joshblack** [comment]
Hey @ashharrison90, I was talking about the conversation with Matt Chapman, sorry about the mistake. I had assumed you two were working together. This was the reason why there were no additional responses in this thread.

I'm sorry if things feel unfair or if there is frustration with this process, I definitely understand. This PR added a ton of great work and I apologize if it doesn't seem like we value it. We really do.

Despite these frustrations, we all need to be respectful to other coworkers. We accept outside contributions as evident in the commit history of this project. I hope you can understand that we also cannot accept all types of contributions from everyone, especially for areas that we ourselves are actively working in or trying to address.

**Pull request closed without merging**
